### ISSUE TITLE
helm/testing: Support for umbrella charts.

### ISSUE THREAD

**Issue by tariq1890** [gruntwork-io/terratest#342]
Thanks for this project, do you have any plans to support unit testing of helm charts which import sub-charts?

**Comment by yorinasub17** [comment]
Hi @tariq1890 , can you elaborate on the use cases of the testing? Specifically, I haven't personally worked with umbrella charts yet, so I am not sure what about the current functions prevent testing them. If there are new functions you think would be useful in testing umbrella charts, knowing those would be helpful as well! We certainly would want to support it if there are deficiencies in the current code!

**Comment by tariq1890** [comment]
In our org, we have a standard helm template chart that is imported in all the charts that we write and maintain.

Unit tests would we very useful if `terratest` can detect these imported charts and render templates successfully.

**Comment by yorinasub17** [comment]
Thanks for those pointers! Perhaps for the short term, we can treat this the same as how terratest handles `terraform init`? That is, expose a manual function that does this (`helm.UpdateDependencies`?) and call that in your test based on what chart you are testing?

**Comment by llibicpep** [comment]
Yeah, that's what I was thinking about. Here's some code I came up with to workaround this for now:

```go
func ensureHelmDependencies(t *testing.T, chartDir string) {
	if !files.FileExists(chartDir) {
		require.FailNow(t, "Chart not found: "+chartDir)
	}

	if !files.FileExists(chartDir + "/requirements.yaml") {
		logger.Logf(t, "Skip chart dependencies - no requirements.yaml")
		return
	}

	if files.FileExists(chartDir + "/charts") {
		logger.Logf(t, "Skip chart dependencies - charts folder already present")
		return
	}

	var buildOrUpdate string
	if files.FileExists(chartDir + "/requirements.lock") {
		buildOrUpdate = "build"
	} else {
		buildOrUpdate = "update"
	}

	if _, err := helm.RunHelmCommandAndGetOutputE(t, &helm.Options{}, "dependency", buildOrUpdate, chartDir); err != nil {
		require.NoError(t, err)
	}
}
```

**Comment by xhu-cloudera** [comment]
Is `requirements.yaml` file necessary? I ran `helm dependency update` and it will create a `charts` folder, with dependencies in tgz files. But I don't see a `requirements.yaml` file. Moreover, right now, the code check for existence of individual template files, however, with dependencies in zip files, that is not the case anymore. What do you guys think the best way to approach this?

**Comment by xhu-cloudera** [comment]
Is someone working on this? I would like to give it a try.

**Comment by xhu-cloudera** [comment]
Here is some thoughts on conditional checking file existence for template files.
```
	for _, templateFile := range templateFiles {
		// validate this is a valid template file
		absTemplateFile := filepath.Join(absChartDir, templateFile)
                // Check file existence only if the template file is from the current chart, not in dependencies
		if !strings.HasPrefix(templateFile, "charts") && !files.FileExists(absTemplateFile) {
			return "", errors.WithStackTrace(TemplateFileNotFoundError{Path: templateFile, ChartDir: absChartDir})
		}

		// Note: we only get the abs template file path to check it actually exists, but the `helm template` command
		// expects the relative path from the chart.
		args = append(args, "--show-only", templateFile)
	}
```

**Comment by aabouzaid** [comment]
Is there any update about this issue?

I have a parent chart with sub-charts and Terratest fails when the parent chart uses templates from the sub-charts ... it's not able to resolve the template names.

I will create a PoC bug report soon.

**Comment by grem11n** [comment]
One can workaround this issue by managing dependencies separately. For example, executing `helm repo add ...` and `helm dep update ...` as a separate step in CI.

However, if dependency charts are packaged with the standard Helm mechanism e.g. as a `.tgz` archive, you cannot use `RenderTemplate()` and `RenderTemplateE()` functions. This [validation step](https://github.com/gruntwork-io/terratest/blob/master/modules/helm/template.go#L53) would fail.

Potentially, be removing that validation step Terratest can allow running templating against those packaged sub-charts. An obvious downside for this is that it has to rely on Helm error messages in this case, which are not always clear enough.

Another possibility is to check if a sub-chart's template is provided there. For example, if a `templateFile` has `charts/` prefix, then skip that check.

For now, one can use [`RunHelmCommandAndGetStdOutE()`](https://github.com/gruntwork-io/terratest/blob/a4eeeaba8a5ee26124cbc27e1d02cc8ebf9e9afa/modules/helm/cmd.go#L60) function to render sub-chart's templates, but you have to re-implement all the helpers of the `RenderTemplateE()` function yourself in this case.

**Comment by aabouzaid** [comment]
OK, just by installing the dependencies before the tests, I was able to [access a dependency from the sub-chart](https://github.com/camunda/camunda-platform-helm/blob/8b5aade10604c443b447b516e01bf8494ff086dc/charts/camunda-platform/templates/ingress.yaml#L23) and [write a test for it](https://github.com/camunda/camunda-platform-helm/blob/8b5aade10604c443b447b516e01bf8494ff086dc/charts/camunda-platform/test/ingress_test.go#L37).

Here is the hierarchy:
> main chart |== has ==> sub-chart |== depends_on ==> external chart

I'm not sure what exactly the goal to achive in the current issue is. Since there is neither `Definition of Done` nor `Acceptance Criteria`.

**Comment by james03160927** [comment]
Closing this issue due to 6 years of inactivity. Feel free to reopen if it still needs to be addressed.
